examples/MomentoRedisLambdaExample/MomentoRedisExampleLambdaHandler/src/MomentoRedisExampleLambdaHandler/MomentoCache.cs

```diff
@@ -0,0 +1,188 @@
+    public async Task<List<Task<GetResult>>> DoMomentoGets(int numMomentoGetRequests, bool compressValue, bool withSleep = false) {
+#pragma warning restore CS1998
+        List<Task<GetResult>> momentoGetTasks = new List<Task<GetResult>>();
+
+        for (int i = 1; i <= numMomentoGetRequests; i++)
+        {
+            Console.WriteLine("Executing MomentoGets");
+            string key = $"key-{i}";
+            momentoGetTasks.Add(ExecuteGet(key, compressValue, withSleep));
+        }
+
+        return momentoGetTasks;
+    }
+
+    private async Task<GetResult> ExecuteGet(string key, bool compressValue, bool withSleep)
+    {
+        var withCompressionStopWatch = new Stopwatch();
+        Console.WriteLine($"Beginning get request for {key}");
+        var stopWatch = Stopwatch.StartNew();
+        if (compressValue)
+        {
+            withCompressionStopWatch = Stopwatch.StartNew();
+        }
+        var getResponseTask = _client.GetAsync(_cacheName, key);
+        Console.WriteLine($"\tIssued get request for {key}");
+        var getResponse = await getResponseTask;
+        Console.WriteLine($"\tDone awaiting get response for {key}");
+        
+        if (withSleep)
+        {
+            await Task.Delay(500);
+            Console.WriteLine($"GET returned from sleeping 500ms for key {key}");
+        }
+        
+        stopWatch.Stop();
+        var responseElapsedMillis = stopWatch.ElapsedMilliseconds;
+        long withDecompressionElapsedMillis = 0;
+        long responseSize = 0;
+        long decompressedResponseSize = 0;
+        if (getResponse is CacheGetResponse.Hit hitResponse)
+        {
+            Console.WriteLine($"\tResponse for {key} is a hit; decompressing");
+            responseSize = hitResponse.ValueByteArray.Length;
+            if (compressValue)
+            {
+                var decompressed = Compression.Decompress(hitResponse.ValueByteArray);
+                Console.WriteLine($"\tResponse finished decompressing for {key}");
+                decompressedResponseSize = decompressed.Length;
+                withCompressionStopWatch.Stop();
+                withDecompressionElapsedMillis = withCompressionStopWatch.ElapsedMilliseconds;
+            }
+            
+        }
+        else
+        {
+            Console.WriteLine(
+                $"\n\nGOT SOMETHING BESIDES A HIT FOR KEY {key}; PROBABLY SHOULDN'T USE THIS DATA\n\n");
+        }
+        
+        var getResult = new GetResult(
+            Response: getResponse,
+            ResponseSize: responseSize,
+            DecompressedResponseSize: decompressedResponseSize,
+            ElapsedMillis: responseElapsedMillis,
+            ElapsedMillisWithDecompression: withDecompressionElapsedMillis
+        );
+        Console.WriteLine($"\tReturning get response for {key}: {getResult}");
+        return getResult;
+    }
+    
+    public async void LogMomentoSetResult(List<Task<SetResult>> momentoSetTasks)
```

Comment:
Since you are only calling this at the very end of your program, and you know that the tasks will all be completed by then, it would be more clear if you realized all the tasks first and then made this function synchronous, with an argument of type `List<SetResult>`.

I think you should be able to do this via `Task.WhenAll`.  We could probably replace your `WaitAll` statements with this:

`IEnumerable<SetResult> setResults = await Task.WhenAll(momentoSetTasks);`

I'm basing that on this post:

https://briancaos.wordpress.com/2021/02/15/c-get-results-from-task-whenall/

General rule of thumb is it is better to have as few async functions as possible. You should only declare a function async if you must do an `await` in it, and usually only if it needs to return a `Task`. If you can find ways to consolidate your `awaits` in your main function(s) so that the helper functions can be synchronous, it will make the code easier to understand.